### ISSUE TITLE
Question about route observation

### ISSUE THREAD

**Issue by zhengzl18** [Sephirex-X/LatentDriver#1]
Hi. I noticed that the route part of the env obs seems to directly come from the original logged trajectory (except that it has been transformed into local coordinate system by `downsampled_elements_transformation`). 

Specifically, I found the following lines in `get_obs()`:
https://github.com/Sephirex-X/LatentDriver/blob/eb720d4a444baa34f5653ccb0c6764f34c812b31/simulator/observation.py#L230-L244

and for the `route` variable, it is directly loaded from some `.npy` files saved in the preprocessing step.

So I further checked the preprocessing codes, according to which the stored routes seemed to be the original `log_trajectory` of sdc:
https://github.com/Sephirex-X/LatentDriver/blob/eb720d4a444baa34f5653ccb0c6764f34c812b31/src/preprocess/waymax_preprocess_utils.py#L200-L207

This kind of inputs is somehow strange to me, as the logged trajectory is itself the expected output. 

Is there something I'm missing here? Please correct me if I'm wrong. Thanks.

**Comment by Sephirex-X** [comment]
> Hi. I noticed that the route part of the env obs seems to directly come from the original logged trajectory (except that it has been transformed into local coordinate system by `downsampled_elements_transformation`).
> 
> Specifically, I found the following lines in `get_obs()`:
> 
> https://github.com/Sephirex-X/LatentDriver/blob/eb720d4a444baa34f5653ccb0c6764f34c812b31/simulator/observation.py#L230-L244
> 
> and for the `route` variable, it is directly loaded from some `.npy` files saved in the preprocessing step.
> 
> So I further checked the preprocessing codes, according to which the stored routes seemed to be the original `log_trajectory` of sdc:
> 
> https://github.com/Sephirex-X/LatentDriver/blob/eb720d4a444baa34f5653ccb0c6764f34c812b31/src/preprocess/waymax_preprocess_utils.py#L200-L207
> 
> This kind of inputs is somehow strange to me, as the logged trajectory is itself the expected output.
> 
> Is there something I'm missing here? Please correct me if I'm wrong. Thanks.

Hi, thank you for your question.  We indeed use the ego-car's logged trajectory to get route observation, but it's reasonable here. 

First, the route for the ego has been absent in waymax for a long time (see [here](https://github.com/waymo-research/waymax/issues/31) and [here](https://github.com/waymo-research/waymax/issues/25)). If you check fig.2(2) on waymax's paper you will see that the route is the union of logged trajectories and drivable future, since the drivable future is not available we use the logged trajectories instead. 

Second, the logged trajectory has been processed and will not leak the ground truth actions here. Specifically, we will unify resample the logged trajectory into a new trajectory and downsample it into object-level representation just like PlanT. You can visualize the model's input to see that. And the code we preprocess route is: https://github.com/Sephirex-X/LatentDriver/blob/eb720d4a444baa34f5653ccb0c6764f34c812b31/src/preprocess/waymax_preprocess_utils.py#L9-L13

Hope this can help you.

**Comment by zhengzl18** [comment]
I see, thanks for clarifying that. I'm bothered by WOMD's lack of drivable futures too. 
Still got a small question though. If constructing route obs requires a certain length of future logged trajectory, then how can closed-loop testing be carried on? For example, suppose the route obs comes from the logged trajectory in the following 8 seconds, then how to get that after one step when the length of the remaining logged trajectory is less than 8 seconds?

**Comment by Sephirex-X** [comment]
> I see, thanks for clarifying that. I'm bothered by WOMD's lack of drivable futures too. Still got a small question though. If constructing route obs requires a certain length of future logged trajectory, then how can closed-loop testing be carried on? For example, suppose the route obs comes from the logged trajectory in the following 8 seconds, then how to get that after one step when the length of the remaining logged trajectory is less than 8 seconds?

Unfortunately, during the simulation, the total length of the route has to be limited by the length of the logged trajectory (you can check our demos, the termination point is fixed). Alternatively, according to the paper, you can perform DFS to get your future trajectories. However, it is extremely hard since we don't know the whole map (WOMO has split the driving record from 20s into several 8s subsequences without a specific id hierarchy). I am afraid all we can do is treat the logged trajectory as the route for now until a better solution has come out.

**Comment by Sephirex-X** [comment]
> > > I see, thanks for clarifying that. I'm bothered by WOMD's lack of drivable futures too. Still got a small question though. If constructing route obs requires a certain length of future logged trajectory, then how can closed-loop testing be carried on? For example, suppose the route obs comes from the logged trajectory in the following 8 seconds, then how to get that after one step when the length of the remaining logged trajectory is less than 8 seconds?
> > 
> > 
> > Unfortunately, during the simulation, the total length of the route has to be limited by the length of the logged trajectory (you can check our demos, the termination point is fixed). Alternatively, according to the paper, you can perform DFS to get your future trajectories. However, it is extremely hard since we don't know the whole map (WOMO has split the driving record from 20s into several 8s subsequences without a specific id hierarchy). I am afraid all we can do is treat the logged trajectory as the route for now until a better solution has come out.
> 
> Yeah I understand that.
> 
> I'm wondering about this: Suppose that for constructing route obs at `t=10`, `state.log_trajectory[..., 11:91]` is required, then during closed-loop testing, how to get route obs at `t=11` when `state.log_trajectory[..., 12:92]` is out of range?
> 
> If route obs at `t=11` is generated from `state.log_trajectory[..., 12:91]`, then wouldn't these route obses at different timesteps have inconsistent meanings?

We treat the whole route as the model's input.  The second dim on `state.log_trajectory` represents `T`, but the processed route doesn't have this dim.

**Comment by zhengzl18** [comment]
I see. Thanks a lot.

**Comment by zhengzl18** [comment]
Hi again, I'd like to ask another question. I noticed that `log_trajectory` is missing in validation and testing tf_examples, which makes route observations inavailable. Did you split the original womd training set into your own training set and validation set? So that route observations are still available in the validation progress.

**Comment by Sephirex-X** [comment]
> Hi again, I'd like to ask another question. I noticed that `log_trajectory` is missing in validation and testing tf_examples, which makes route observations inavailable. Did you split the original womd training set into your own training set and validation set? So that route observations are still available in the validation progress.

Hi, we read this directly from the tf_example, you may check your data first.

**Comment by zhengzl18** [comment]
> > Hi again, I'd like to ask another question. I noticed that `log_trajectory` is missing in validation and testing tf_examples, which makes route observations inavailable. Did you split the original womd training set into your own training set and validation set? So that route observations are still available in the validation progress.
> 
> Hi, we read this directly from the tf_example, you may check your data first.

Sorry, I made a mistake. Only test dataset doesn't contain log_trajectory. Thanks for your time!
